tests/run-make/linker-warning/rmake.rs

```diff
@@ -25,4 +44,33 @@ fn main() {
         .assert_stderr_contains("object files omitted")
         .assert_stderr_contains_regex(r"\{")
         .assert_stderr_not_contains_regex(r"lib(/|\\\\)libstd");
+
+    // Make sure we show linker warnings even across `-Z no-link`
+    rustc()
+        .arg("-Zno-link")
+        .input("-")
+        .stdin_buf("#![deny(linker_messages)] \n fn main() {}")
+        .run()
+        .assert_stderr_equals("");
+    rustc()
+        .arg("-Zlink-only")
+        .arg("rust_out.rlink")
+        .linker("./fake-linker")
+        .link_arg("run_make_warn")
+        .run_fail()
+        // NOTE: the error message here is quite bad (we don't have a source
+        // span, but still try to print the lint source). But `-Z link-only` is
+        // unstable and this still shows the linker warning itself so this is
+        // probably good enough.
+        .assert_stderr_contains("linker stderr: bar");
+
+    // Same thing, but with json output. This currently isn't supported and just ICEs.
```

Comment:
Can you at least add a FIXME at the location where this ICEs?

Reply:
Actually, this will have to be fixed before the PR can land as cargo always uses the json output format and thus would cause an ICE on any linker warning.

Reply:
I have fixed this to use a dummy span whenever a source map is missing. I don't think it should happen except when all three of -Z no-link, --error-format=json, and a linker_messages lint are present, but i agree it's better not to have panics floating around for no reason.